### PR TITLE
Best practices for libpacemaker

### PR DESCRIPTION
**Pull request by kgaillot**
This continues the project to bring libpacemaker up to current development guidelines, finishing pcmk_sched_instances.c and starting pcmk_sched_clone.c.

### CODE REVIEW

**Comment by kgaillot** [comment]
@nrwahl2 , can you review this when you get a chance? Low priority

File: lib/pacemaker/pcmk_sched_clone.c [hunk 1]
```diff
@@ -39,39 +39,45 @@ add_dependent_scores(gpointer data, gpointer user_data)
...
- * \return Node that \p rsc is assigned to, if assigned entirely to one node
+ * \return NULL (clones are not assigned to a single node)
  */
 pe_node_t *
-pcmk__clone_allocate(pe_resource_t *rsc, const pe_node_t *prefer)
+pcmk__clone_assign(pe_resource_t *rsc, const pe_node_t *prefer)
 {
+    GList *iter = NULL;
+
+    CRM_ASSERT((rsc != NULL) && (rsc->variant == pe_clone));
```
Comment (nrwahl2) [inline lib/pacemaker/pcmk_sched_clone.c:54]: Could use `pe_rsc_is_clone(rsc)` if you want

File: lib/pacemaker/pcmk_sched_clone.c [hunk 2]
```diff
@@ -39,39 +39,45 @@ add_dependent_scores(gpointer data, gpointer user_data)
...
  * \param[in]     prefer  Node to prefer, if all else is equal
  *
- * \return Node that \p rsc is assigned to, if assigned entirely to one node
+ * \return NULL (clones are not assigned to a single node)
  */
 pe_node_t *
-pcmk__clone_allocate(pe_resource_t *rsc, const pe_node_t *prefer)
+pcmk__clone_assign(pe_resource_t *rsc, const pe_node_t *prefer)
 {
+    GList *iter = NULL;
```
Comment (nrwahl2) [inline lib/pacemaker/pcmk_sched_clone.c:52]: We only use this once if you wanted to keep it in the loop header. Not worth bothering with unless you're already changing something else.

**Review by nrwahl2 — COMMENTED** [review]
Nice

File: lib/pacemaker/pcmk_sched_clone.c (hunk 1 above)
Comment (kgaillot) [reply lib/pacemaker/pcmk_sched_clone.c:54]: definitely

File: lib/pacemaker/pcmk_sched_clone.c (hunk 2 above)
Comment (kgaillot) [reply lib/pacemaker/pcmk_sched_clone.c:52]: thanks

**Comment by kgaillot** [comment]
Added a couple commits on top to address review

File: lib/pengine/clone.c [hunk 3]
```diff
@@ -1371,3 +1371,22 @@ pe__create_promotable_pseudo_ops(pe_resource_t *clone, bool any_promoting,
...
 }
+
+/*!
+ * \internal
+ * \brief Create all notification data and actions for a clone
+ *
+ * \param[in,out] clone  Clone to create notifications for
+ */
+void
+pe__create_clone_notifications(pe_resource_t *clone)
```
[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/pengine/clone.c:1399]: Sounds good :)

Same comments apply to the free functions in the next commit

arguably for `pe__create_clone_notif_pseudo_ops()` too

[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/pengine/clone.c:1399]: oh right, meant to do that and forgot -- will make a note for later

**Pull request merged**